### ISSUE TITLE
how to handle template dependencies

### ISSUE THREAD

**Issue by oba11** [kelseyhightower/confd#292]
I have two templates with one depending on the other one before the actual service runs.
Currently, I get an error that the file is missing.
Is there any way I can handle this dependency?

Thanks.

**Comment by kelseyhightower** [comment]
@oba11 Currently confd does not support dependencies between templates, I'm not sure if this kind of feature is in scope for confd. 

Can you tell me more about what you are trying to accomplish?

**Comment by oba11** [comment]
thanks @kelseyhightower for the response.
Basically, I have 2 nginx templates (`/etc/nginx/sites-enabled/default.conf` and `/etc/nginx/conf.d/custom-location.conf`) where one includes the other one.

/etc/nginx/sites-enabled/default.conf
```
server {
  server_name         hello.{{$nginxDomain}};
  listen              80;
  listen              443 spdy ssl;
  include             /etc/nginx/conf.d/custom-location.conf;
}
```

`default.conf` gets generated by confd before `custom-location.conf` where nginx throws an error that `custom-location.conf` is not found.

**Comment by joshk0** [comment]
I have this problem too. We use confd to propagate a haproxy configuration as well as a SSL certificate. The haproxy configuration references the SSL certificate. If the SSL certificate comes first, it all works great. If the haproxy.cfg comes first, the config check fails because the SSL certificate is not yet present.

I think you wouldn't really need to do anything except ensure that templates are rendered in lexical order when a whole batch is received at once.

How about it?

**Comment by bacongobbler** [comment]
@joshk0 maybe the resolution is to run two separate `confd` processes: one for each stage in the deployment cycle. I don't know if adding a lexical ordering to `confd` could be reasonably supported, but PRs proposing a solution would be appreciated :)

**Comment by Dynom** [comment]
I currently solve this by using the poor-man solution:

    001-initial-template
    002-followup-template
    003-another-template-that-depends-on-initial-template

**Comment by oba11** [comment]
Thanks @Dynom would try it out soon and get back to you. 
Thanks again.

**Comment by joshk0** [comment]
So wait. There is already a functional lexical ordering?

**Comment by bacongobbler** [comment]
Well theoretically we could change confd's behaviour to run all templates in parallel, but it sounds like this works as a way to run templates in a lexical ordering for now. I'd recommend using template names like `0001-sometemplate` to achieve the desired effect.

**Comment by hawka** [comment]
Lexicographical ordering of templates is not resolving this issue for me. We are seeing an issue wherein the checks performed in the second template are not able to succeed due to the first template not writing it's finalized configuration out.

See below logs:
```
Aug 26 16:02:02 harrier confd[18]: DEBUG Loading template resources from confdir /etc/confd
Aug 26 16:02:02 harrier confd[18]: DEBUG Loading template resource from /etc/confd/conf.d/001_ssl_cert.toml
Aug 26 16:02:02 harrier confd[18]: DEBUG Loading template resource from /etc/confd/conf.d/002_haproxy.toml
Aug 26 16:02:02 harrier confd[18]: DEBUG Retrieving keys from store
Aug 26 16:02:02 harrier confd[18]: DEBUG Key prefix set to /
Aug 26 16:02:02 harrier confd[18]: DEBUG Retrieving keys from store
Aug 26 16:02:02 harrier confd[18]: DEBUG Key prefix set to /
Aug 26 16:02:02 harrier confd[18]: DEBUG Using source template /etc/confd/templates/001_ssl_cert.tmpl
Aug 26 16:02:02 harrier confd[18]: DEBUG Compiling source template /etc/confd/templates/001_ssl_cert.tmpl
Aug 26 16:02:02 harrier confd[18]: INFO Keeping staged file: /var/tmp/.ssl_cert.pem122139504
Aug 26 16:02:02 harrier confd[18]: DEBUG Comparing candidate config to /var/tmp/ssl_cert.pem
Aug 26 16:02:02 harrier confd[18]: INFO Target config /var/tmp/ssl_cert.pem out of sync
Aug 26 16:02:02 harrier confd[18]: DEBUG Running openssl verify -CAfile /etc/ssl/certs/ca-certificates.crt -untrusted /var/tmp/.ssl_cert.pem122139504 /var/tmp/.s
Aug 26 16:02:02 harrier confd[18]: DEBUG Using source template /etc/confd/templates/002_haproxy.tmpl
Aug 26 16:02:02 harrier confd[18]: DEBUG Compiling source template /etc/confd/templates/002_haproxy.tmpl
Aug 26 16:02:02 harrier confd[18]: INFO Keeping staged file: /var/tmp/.haproxy.cfg368877583
Aug 26 16:02:02 harrier confd[18]: DEBUG Comparing candidate config to /var/tmp/haproxy.cfg
Aug 26 16:02:02 harrier confd[18]: INFO Target config /var/tmp/haproxy.cfg out of sync
Aug 26 16:02:02 harrier confd[18]: DEBUG Running /usr/sbin/haproxy -c -f /var/tmp/.haproxy.cfg368877583
Aug 26 16:02:02 harrier confd[18]: ERROR Config check failed: exit status 1
Aug 26 16:02:02 harrier confd[18]: DEBUG "/var/tmp/.ssl_cert.pem122139504: OK\n"
Aug 26 16:02:02 harrier confd[18]: DEBUG Overwriting target config /var/tmp/ssl_cert.pem
Aug 26 16:02:02 harrier confd[18]: DEBUG Running /usr/sbin/haproxy -f /var/tmp/haproxy.cfg -D -st $(cat /var/tmp/haproxy.pid)
Aug 26 16:02:02 harrier confd[18]: ERROR exit status 1
```

I posit that confd is doing a sweep of all the templates it wants to evaluate that have changed, and then running the checks, and then updating the target files if it succeeds. However, while some of the checks (i.e. openssl verify) run against the temporary file, haproxy.cfg expects to see the ssl_cert.pem file. This occurs before the first template's output is written from the temporary file to the actual one.

**Comment by oba11** [comment]
Tried the lexical ordering. Sometimes it works and other times, it doesn't. The template file was carefully arranged in order 

```
001-
002-
003-
where main config is 003-
```

Sometimes it follows the pattern, other times it generates them at random.

**Comment by Dynom** [comment]
This approach works flawlessly for me. However I only use 3 files at most, where only two have a lexical dependency. Maybe if I add more it becomes a problem. 

I think this proves that we really need a system to handle dependencies reliably, even if it is a simple sorting strategy when reading the template files.

**Comment by hawka** [comment]
I noticed that the [util_test.go file is doing a sort](https://github.com/kelseyhightower/confd/blob/master/src/github.com/kelseyhightower/confd/resource/template/util_test.go#L111) on the strings returned by the recursive file find -- it would be interesting to take that sort out, and see if the test still succeeds on all platforms.

**Comment by Dynom** [comment]
Interesting find @hawka, that might actually be why that test is never failing on that part.

@oba11 on what fs are you testing it? I've read up on some of the default ordering characteristics and apparently the order where you create the files in, matters for functions such as readfile. I'm curious why you get mixed results, that doesn't make sense to me.

**Comment by oba11** [comment]
was testing it using aufs (mainly `alpine` distro from docker).

**Comment by dverbeek84** [comment]
@oba11 same issue here ordering with 001, 002 is not working.
It is execute random. My nginx config is depending on cert en key(witch i create with confd). But sometimes nginx config is created before the keys and with the watch options this is not good :(

**Comment by schuylr** [comment]
Lexical ordering does not work for me either. Also tried ensuring that the files are created in a particular order with no success. I'm going to open a new issue about this.
